assets/js/base/hooks/cart/use-store-cart-coupons.js

```diff
@@ -56,9 +62,9 @@ export const useStoreCartCoupons = () => {
 						}
 					} )
 					.catch( ( error ) => {
-						addErrorNotice( error.message, {
-							id: 'coupon-form',
-						} );
+						setValidationErrors( { coupon: error.message } );
+						// Finished handling the coupon.
+						receiveApplyingCoupon( '' );
```

Comment:
I needed to move setting the loading state for `applyingCoupon` outside of the `applyCoupon` action (for errors only) because throwing an error _after_ setting the loading state in the action meant that validation errors were set _after_ the the loading is `false`.  In turn this meant that in the `TotalsCouponCodeInput` logic, the invalid coupon code would always be cleared from the input.

I tried various different approaches (including using the cartData error state) but this seemed to be the cleanest. The downside is any code catching errors will _have_ to explicitly cancel the `receiveApplyingCoupon` error state. I think that's okay because `useStoreCartCoupons` is generally the interface components will go through anyways.